src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java

```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * Gets the header associated with our driving source of variants as a VCFHeader.
+     *
+     * @return VCFHeader for our driving source of variants
+     */
+    public final VCFHeader getHeaderForVariants() {
+        final Object header = drivingVariants.getHeader();
+
+        if ( ! (header instanceof VCFHeader) ) {
+            throw new GATKException("Header for " + drivingVariantFile + " is not in VCF header format");
+        }
+
+        return (VCFHeader)header;
+    }
+
+    @Override
+    public boolean requiresReference() {
+        return true;
+    }
+
+    /**
+     * Implementation of variant-based traversal.
+     * Subclasses can override to provide their own behavior but default implementation should be suitable for most uses.
+     */
+    @Override
+    public void traverse() {
+        final CountingReadFilter readFilter = makeReadFilter();
+        final VariantFilter variantFilter = makeVariantFilter();
+
+        if (traverseByLocus) {
+            getLocusStream(getTraversalIntervals())
+                .forEach(locus -> {
+                    // Get all of the variants overlapping this locus, filter them and collect into a list
+                    final Iterator<VariantContext> overlappingVariants = drivingVariants.query(locus);
+                    final List<VariantContext> filteredVariants = getTransformedVariantStream(
+                            Spliterators.spliteratorUnknownSize(overlappingVariants, 0), variantFilter)
+                            .collect(Collectors.toList());
+                        if (!filteredVariants.isEmpty()) {
+                            apply(locus,
+                                    filteredVariants,
+                                    new ReadsContext(reads, locus, readFilter),
+                                    new ReferenceContext(reference, locus),
+                                    new FeatureContext(features, locus));
+
+                            progressMeter.update(locus);
+                        }
+                    }
+            );
+        } else {
+            // Process each variant in the input stream.
+            getTransformedVariantStream( getSpliteratorForDrivingVariants(), variantFilter )
+                    .forEach(variant -> {
+                        final SimpleInterval variantInterval = new SimpleInterval(variant);
+                        apply(variant,
+                                Collections.singletonList(variant),
+                                new ReadsContext(reads, variantInterval, readFilter),
+                                new ReferenceContext(reference, variantInterval),
+                                new FeatureContext(features, variantInterval));
+
+                        progressMeter.update(variantInterval);
+                    });
+        }
+    }
+
+    protected Stream<VariantContext> getTransformedVariantStream(final Spliterator<VariantContext> source, final VariantFilter filter) {
```

Comment:
This seems like it could maybe be pulled up to the base class and avoid duplicating it here and in multi and single variant walkers.

Reply:
Yes, done. It still takes a spliterator, though..